diff --git a/block/blkdebug.c b/block/blkdebug.c
index dee3a0e..b00d0f0 100644
--- a/block/blkdebug.c
+++ b/block/blkdebug.c
@@ -954,7 +954,7 @@ static void blkdebug_refresh_filename(BlockDriverState *bs)
     }
 
     ret = snprintf(bs->exact_filename, sizeof(bs->exact_filename),
-                   "blkdebug:%s:%s",
+                   "blkdebug:%s:%.4085s",
                    s->config_file ?: "", bs->file->bs->exact_filename);
     if (ret >= sizeof(bs->exact_filename)) {
         /* An overflow makes the filename unusable, so do not report any */
diff --git a/block/blkverify.c b/block/blkverify.c
index c5f8e8d..b242680 100644
--- a/block/blkverify.c
+++ b/block/blkverify.c
@@ -290,7 +290,7 @@ static void blkverify_refresh_filename(BlockDriverState *bs)
         && s->test_file->bs->exact_filename[0])
     {
         int ret = snprintf(bs->exact_filename, sizeof(bs->exact_filename),
-                           "blkverify:%s:%s",
+                           "blkverify:%.2000s:%.2000s",
                            bs->file->bs->exact_filename,
                            s->test_file->bs->exact_filename);
         if (ret >= sizeof(bs->exact_filename)) {
diff --git a/hw/usb/bus.c b/hw/usb/bus.c
index ee6b43a..c8a5f83 100644
--- a/hw/usb/bus.c
+++ b/hw/usb/bus.c
@@ -401,7 +401,7 @@ void usb_register_companion(const char *masterbus, USBPort *ports[],
 void usb_port_location(USBPort *downstream, USBPort *upstream, int portnr)
 {
     if (upstream) {
-        int l = snprintf(downstream->path, sizeof(downstream->path), "%s.%d",
+        int l = snprintf(downstream->path, sizeof(downstream->path), "%.12s.%d",
                          upstream->path, portnr);
         /* Max string is nn.nn.nn.nn.nn, which fits in 16 bytes */
         assert(l < sizeof(downstream->path));
