diff -r 9f058bee9211 -r 67215db1e6ce run_ml.sh
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/run_ml.sh	Mon Apr 08 00:20:56 2013 +0300
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+QEMU_PATH=qemu-1.4.0
+ML_PATH=../magic-lantern/platform/$1.$2
+
+make -C $QEMU_PATH || exit
+make -C $ML_PATH || exit
+make qemu-helper.bin -C $ML_PATH || exit
+cp $ML_PATH/autoexec.bin .
+cp $ML_PATH/qemu-helper.bin .
+
+rm vram.txt
+rm vram.png
+
+$QEMU_PATH/arm-softmmu/qemu-system-arm -M ML-$1
+
+convert vram.txt vram.png && eog vram.png
diff -r 9f058bee9211 -r 67215db1e6ce run_ml_500D.sh
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/run_ml_500D.sh	Mon Apr 08 00:20:56 2013 +0300
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sh run_ml.sh 500D 111
diff -r 9f058bee9211 -r 67215db1e6ce run_ml_5D2.sh
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/run_ml_5D2.sh	Mon Apr 08 00:20:56 2013 +0300
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sh run_ml.sh 5D2 212
diff -r 9f058bee9211 -r 67215db1e6ce run_ml_5D3.sh
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/run_ml_5D3.sh	Mon Apr 08 00:20:56 2013 +0300
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sh run_ml.sh 5D3 113
diff -r 9f058bee9211 -r 67215db1e6ce run_ml_600D.sh
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/run_ml_600D.sh	Mon Apr 08 00:20:56 2013 +0300
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sh run_ml.sh 600D 102
diff -r 9f058bee9211 -r 67215db1e6ce run_ml_60D.sh
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/run_ml_60D.sh	Mon Apr 08 00:20:56 2013 +0300
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sh run_ml.sh 60D 111
diff -r 9f058bee9211 -r 67215db1e6ce run_ml_650D.sh
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/run_ml_650D.sh	Mon Apr 08 00:20:56 2013 +0300
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+sh run_ml.sh 650D 101
